### ISSUE TITLE
when counting security group iprules, other security groups added to them are being counted as one rule

### ISSUE THREAD

**Issue by rpenugonda** [jantman/awslimitchecker#431]
Issue:

we have security groups which include other security groups for ingress via kubernetes. When i run awslimitchecker it only counts the `len(sg.ip_permissions)`. all the inserted security groups are in one rule like below and you get the *EC2/Rules per VPC security group  * count wrong (i think)
`{'IpProtocol': '-1', 'IpRanges': [], 'Ipv6Ranges': [], 'PrefixListIds': [], 'UserIdGroupPairs': [{'GroupId': 'sg-00375f', 'UserId': 'xxx'}, {'GroupId': 'sg-00375f', 'UserId': 'xxx'}, ...]`

so i i have 20 groups added to a security group inbound i get the rule count as 1. this below fixed our code base
awslimitchecker/services/ec2.py
```
445     def _find_usage_networking_sgs(self):
446         """calculate usage for VPC-related things"""
447         logger.debug("Getting usage for EC2 VPC resources")
448         sgs_per_vpc = defaultdict(int)
449         rules_per_sg = defaultdict(int)
450         for sg in self.resource_conn.security_groups.all():
451             if sg.vpc_id is not None:
452                 sgs_per_vpc[sg.vpc_id] += 1
453             count = 1
454             if sg.vpc_id is not None:
455                 for rule in sg.ip_permissions:
456                     if len(rule['UserIdGroupPairs']) == 0:
457                         rules_per_sg[sg.id] = 1 + rules_per_sg[sg.id]
458                     else:
459                         rules_per_sg[sg.id] = rules_per_sg[sg.id] + len(rule['UserIdGroupPairs'])
460                    #rules_per_sg[sg.id] = len(sg.ip_permissions)
461 
```

**Comment by rpenugonda** [comment]
awslimitchecker --version
awslimitchecker 7.1.0 is AGPL-licensed free software; all users have a right to the full source code of this version. See <https://github.com/jantman/awslimitchecker>
awslimitchecker 7.1.0 (see <https://github.com/jantman/awslimitchecker> for source code)

**Comment by jantman** [comment]
@rpenugonda have you confirmed with AWS how they evaluate this limit, or done any experimentation to confirm it? The section of code in question hasn't been changed in over four years, so while it's possible that it's been wrong all this time, it doesn't seem too likely.

This doesn't necessarily mean that you're wrong, the [documentation for the SecurityGroup resource](https://docs.aws.amazon.com/AWSEC2/latest/APIReference/API_SecurityGroup.html) returned by the ``DescribeSecurityGroups`` API call lists ``ipPermissions`` as

> The inbound rules associated with the security group.

This would imply to me that each element in the ``ip_permissions`` list is a "rule" (as counted towards limits), and the current code is correct.

**Comment by bazzargh** [comment]
@jantman I work with @rpenugonda - we'd hit the rules-per-sg limit with these and it's easy to reproduce; in one of our accounts with the default rules-per-sg limit of 60, we did:
```
for num in $(seq 1 70); do aws --region eu-west-1 ec2 create-security-group --vpc-id $vpcid --description "brian test $num" --group-name brian-test-$num >> groups; done
for sgid in $(grep -oE 'sg-[0-9a-f]*' groups); do aws --region eu-west-1 ec2 authorize-security-group-ingress --protocol tcp --port 80 --group-id $targetsgid --source-group $sgid; done
```

The last 10 attempted adds all errored:
```
An error occurred (RulesPerSecurityGroupLimitExceeded) when calling the AuthorizeSecurityGroupIngress operation: The maximum number of rules per security group has been reached.
exit status 255
```

Testing with the code in limitchecker:
```
>>> len(client.SecurityGroup('<sgid>').ip_permissions)
1
```

The contents of that 1 row was:
```
[{'FromPort': 80, 'IpProtocol': 'tcp', 'IpRanges': [], 'Ipv6Ranges': [], 'PrefixListIds': [], 'ToPort': 80, 'UserIdGroupPairs': [{'GroupId': 'sg-0057c963d51a18402', 'UserId': '<acct>'}, ... 59 more]}]
```
So yes, despite appearing to be consolidated in a single rule in the api, these are being counted separately.

**Comment by patrickpierson** [comment]
Seeing this issue as well.  Looks like rules are counted by port in my case.  If I have two entries on the same port but with different CIDRs it is counted as one rule via awslimitchecker.  AWS counts it as two rules.

**Comment by patrickpierson** [comment]
> @rpenugonda have you confirmed with AWS how they evaluate this limit, or done any experimentation to confirm it? The section of code in question hasn't been changed in over four years, so while it's possible that it's been wrong all this time, it doesn't seem too likely.
> 
> This doesn't necessarily mean that you're wrong, the [documentation for the SecurityGroup resource](https://docs.aws.amazon.com/AWSEC2/latest/APIReference/API_SecurityGroup.html) returned by the `DescribeSecurityGroups` API call lists `ipPermissions` as
> 
> > The inbound rules associated with the security group.
> 
> This would imply to me that each element in the `ip_permissions` list is a "rule" (as counted towards limits), and the current code is correct.

Just to follow up on my comment above.  AWS evaluates security group rules by the number of entries. It does not lump all of the CIDRs for a specific port into one rule. Each entry displayed in the console is one rule.

**Comment by jantman** [comment]
Ok, thanks for all of the information on this. I'm on vacation this week, but I'll dig into a fix for this next week when I'm back. Apologies for the delay.

**Comment by jantman** [comment]
Many apologies for the delay in implementing this... it's occasionally difficult to balance life and maintaining an open source project in my spare time. I plan on trying to get a fix for this out in the next week.

**Comment by jantman** [comment]
@rpenugonda @bazzargh @patrickpierson I'm hoping to get a release out in the next day or so and this is one of only 2 remaining issues, so I've been doing some work on this today.

I've followed up on [bazzargh's example](https://github.com/jantman/awslimitchecker/issues/431#issuecomment-532741053) and decided to do some more in-depth testing with some generated data, working in the awslimitchecker integration test account which has its Rules per VPC Security Group limit set to the default of 60. The results have been rather enlightening, and somewhat obvious in hindsight from a networking perspective...

* Yes, my understanding (and awslimitchecker's calculation) was incorrect... even though the API calls them "rules", that's definitely not how the limit is evaluated.
* Yes, as shown by @bazzargh's testing, if you create an SG and attempt to authorize ingress or egress for a given port/protocol, you can add up 60 ``IpRanges`` or 60 ``Ipv6Ranges`` or 60 ``UserIdGroupPairs`` and will get a ``RulesPerSecurityGroupLimitExceeded`` exception when trying to add the 61st.
* **however**, when you mix multiple types of IpPermissions (`IpRanges``, ``Ipv6Ranges``, ``UserIdGroupPairs``, ``PrefixListIds``) it becomes obvious that behind the scenes, this limit is just a limit of firewall rules

So in reality, after about 100 iterations of testing, the result is that each Security Group has a maximum of four rule sets:
* IPv4 Ingress
* IPv6 Ingress
* IPv4 Egress
* IPv6 Egress

Each of these rule sets can handle a maximum of 60 "permissions", where a permission is _either_ a ``IpRanges`` / ``Ipv6Ranges``, a ``UserIdGroupPairs``, or a ``PrefixListIds``.

Essentially, you stop being able to add permissions to either ingress or egress (they're evaluated separately; you can have 60 ingress permissions and 60 egress permissions) when the sum of ``UserIdGroupPairs`` on all rules, plus the sum of ``PrefixListIds`` on all rules, plus the maximum of (the sum of ``IpRanges`` on all rules or the sum of ``Ipv6Ranges`` on all rules) hits 60.

I'll work on updating awslimitchecker accordingly, and should have a branch to test within the next few hours.

**Comment by jantman** [comment]
This is fixed with #447 which is now merged into the ``develop`` branch, if anyone would like to verify the fix. This should be released within the next few days.

**Comment by jantman** [comment]
This has been fixed in [8.0.0](https://github.com/jantman/awslimitchecker/releases/tag/8.0.0) which is now live on [PyPI](https://pypi.org/project/awslimitchecker/) and [Docker Hub](https://hub.docker.com/r/jantman/awslimitchecker/tags). Apologies for any delays in getting this released.

**Comment by bergkampsliew** [comment]
sorry for reading this issue late. 
can i ask - if the latest fix/alert is based on total 120 ? https://docs.aws.amazon.com/vpc/latest/userguide/amazon-vpc-limits.html#vpc-limits-security-groups
i believe the current default value is 50 (in awslimitchecker) too.

as i read the awslimitchecker documentation which says - The limit that we alert on is the maximum of those values for ingress and egress.

**Comment by jantman** [comment]
@bergkampsliew You're right, it appears we still have the default set at 50, but it seems to now be 60 per the AWS documentation. I'll open an issue for this and try to get it out in the next release.

**Comment by jantman** [comment]
This has been released in [8.0.1](https://github.com/jantman/awslimitchecker/releases/tag/8.0.1) and is now [live on PyPI](https://pypi.org/project/awslimitchecker/). Thank you so much, and apologies for the delay.
